platform/BufferedSerial.h

```diff
@@ -0,0 +1,94 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_BUFFEREDSERIAL_H
+#define MBED_BUFFEREDSERIAL_H
+
+#include "platform/platform.h"
+
+#if DEVICE_SERIAL
+
+#include "FileHandle.h"
+#include "SerialBase.h"
+#include "PlatformMutex.h"
+#include "serial_api.h"
+#include "CircularBuffer.h"
+
+namespace mbed {
+class BufferedSerial : private SerialBase, public FileHandle {
```

Comment:
The original view plan when we started, that I recall we agreed on, was the the external ATParser and BufferedSerial weren't good enough to go into the core, due to the flaws in FileHandle and the serial-specific nature of ATParser. The system was inferior to the Ublox equivalent in the unsupported/cellular nature.

This whole PR revises the entire ATParser/FileHandle/BufferedSerial stack to work as well as that Ublox code, making FileHandle actually useable for devices for the first time (Serial was useless, and BufferedSerial wasn't a FileHandle).

BufferedSerial is roughly compatible with that external one from an application point of view, but not totally. ATParser is virtually compatible once constructed - we might be able to give it some built-in serial knowledge so it can have a constructor that creates a BufferedSerial if given pin names?

But they're not going to be totally compatible. The only way to totally avoid issues is to bring them in under a new name.

I believe we've totally defeated namespaces with `using namespace mbed` in mbed.h, so making them `mbed::ATParser` doesn't help. (And were the external ones in namespace mbed anyway)?

It really does feel like quite a strong requirement - that any external code being pulled in has to be renamed in case anyone is using it already. That seems to be presupposing that they carry on using the external version, no? Isn't the point that anyone using the external code should be switching to the mbed OS version, once it's pulled in.